flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMailboxTestHarness.java

```diff
@@ -143,12 +143,17 @@ public void waitForTaskCompletion() throws Exception {
 		}
 	}
 
-	@Override
-	public void close() throws Exception {
-		streamTask.cancel();
-
+	public void finishProcessing() throws Exception {
```

Comment:
It's a bit strange that we replicate the shutdown/cancel logic in the TestHarness and don't re-use the production logic. Just noticing it here, it was like this before the PR.

Reply:
Yes, I also found it problematic.

For a time I thought about not using this harness, but the deprecated StreamTaskTestHarness instead which does simply call the invoke method, because of the reason you described. However, because it does not let you process elements step wise, you cannot e.g. issue `cancel` before it finishes.